discovery/uyuni/uyuni.go

```diff
@@ -0,0 +1,327 @@
+	apiURL := &url.URL{
+		Scheme: conf.Host.Scheme,
+		Host:   conf.Host.Host,
+		Path:   uyuniXMLRPCAPIPath,
+	}
+
+	rt, err := config.NewRoundTripperFromConfig(conf.HTTPClientConfig, "uyuni_sd", config.WithHTTP2Disabled())
+	if err != nil {
+		return nil, err
+	}
+
+	d := &Discovery{
+		apiURL:       apiURL,
+		roundTripper: rt,
+		username:     conf.Username,
+		password:     string(conf.Password),
+		entitlement:  conf.Entitlement,
+		separator:    conf.Separator,
+		interval:     time.Duration(conf.RefreshInterval),
+		logger:       logger,
+	}
+	d.Discovery = refresh.NewDiscovery(
+		logger,
+		"uyuni",
+		time.Duration(conf.RefreshInterval),
+		d.refresh,
+	)
+	return d, nil
+}
+
+func (d *Discovery) getEndpointLabels(
+	endpoint endpointInfo,
+	systemGroupIDs []systemGroupID,
+	networkInfo networkInfo,
+) model.LabelSet {
+
+	var addr string
+	managedGroupNames := getSystemGroupNames(systemGroupIDs)
+	addr = fmt.Sprintf("%s:%d", networkInfo.Hostname, endpoint.Port)
+
+	result := model.LabelSet{
+		model.AddressLabel:       model.LabelValue(addr),
+		uyuniLabelMinionHostname: model.LabelValue(networkInfo.Hostname),
+		uyuniLabelPrimaryFQDN:    model.LabelValue(networkInfo.PrimaryFQDN),
+		uyuniLablelSystemID:      model.LabelValue(fmt.Sprintf("%d", endpoint.SystemID)),
+		uyuniLablelGroups:        model.LabelValue(strings.Join(managedGroupNames, d.separator)),
+		uyuniLablelEndpointName:  model.LabelValue(endpoint.EndpointName),
+		uyuniLablelExporter:      model.LabelValue(endpoint.ExporterName),
+		uyuniLabelProxyModule:    model.LabelValue(endpoint.Module),
+		uyuniLabelMetricsPath:    model.LabelValue(endpoint.Path),
+	}
+
+	return result
+}
+
+func getSystemGroupNames(systemGroupsIDs []systemGroupID) []string {
+	managedGroupNames := make([]string, 0, len(systemGroupsIDs))
+	for _, systemGroupInfo := range systemGroupsIDs {
+		managedGroupNames = append(managedGroupNames, systemGroupInfo.GroupName)
+	}
+
+	return managedGroupNames
+}
+
+func (d *Discovery) getTargetsForSystems(
+	rpcClient *xmlrpc.Client,
+	token string,
+	entitlement string,
+) ([]model.LabelSet, error) {
+
+	result := make([]model.LabelSet, 0)
+
+	systemGroupIDsBySystemID, err := getSystemGroupsInfoOfMonitoredClients(rpcClient, token, entitlement)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to get the managed system groups information of monitored clients")
+	}
+
+	systemIDs := make([]int, 0, len(systemGroupIDsBySystemID))
+	for systemID := range systemGroupIDsBySystemID {
+		systemIDs = append(systemIDs, systemID)
+	}
+
+	endpointInfos, err := getEndpointInfoForSystems(rpcClient, token, systemIDs)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to get endpoints information")
+	}
+
+	networkInfoBySystemID, err := getNetworkInformationForSystems(rpcClient, token, systemIDs)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to get the systems network information")
+	}
+
+	for _, endpoint := range endpointInfos {
```

Comment:
somehow I can only get a single endpoint (see the test system, where I have enabled 3 exporters but only 1 is returned). Is that a bug in uyuni?

Reply:
The next question I have about uyuni endpoint: how do I discover the prometheus instances? How do I discover the alertmanager instances? how do I discover the exporter_exporter instances (for monitoring them)?

Reply:
> somehow I can only get a single endpoint (see the test system, where I have enabled 3 exporters but only 1 is returned). Is that a bug in uyuni?

The _prometheus-formula_ does not include relabeling configuration yet. When applying the highstate you have reset the Prometheus configuration as well. I've manually added relabeling configuration again and see three targets now.

Reply:
> The next question I have about uyuni endpoint: how do I discover the prometheus instances? How do I discover the alertmanager instances? how do I discover the exporter_exporter instances (for monitoring them)?

It will have to be added to Uyuni code. The new API call `system.monitoring.listEndpoints` should return all installed application endpoints. It's not implemented yet, currently only exporter endpoints are returned.